pkg/controller/storagecluster/storagecluster.go

```diff
@@ -334,6 +339,69 @@ func (c *Controller) validateCustomAnnotations(current *corev1.StorageCluster) e
 	return nil
 }
 
+func getKeysFromNodeSelector(ns *corev1.NodeSelector) map[string]bool {
+	keys := make(map[string]bool)
+
+	if ns.LabelSelector != nil {
+		for k := range ns.LabelSelector.MatchLabels {
+			keys[k] = true
+		}
+
+		for _, r := range ns.LabelSelector.MatchExpressions {
+			keys[r.Key] = true
+		}
+	}
+
+	return keys
+}
+
+func (c *Controller) validateCloudStorageLabelKey(cluster *corev1.StorageCluster) error {
+	key, err := c.getCloudStorageLabelKey(cluster)
+	if err != nil {
+		return err
+	}
+
+	if cluster.Spec.CloudStorage != nil &&
+		cluster.Spec.CloudStorage.NodePoolLabel != "" &&
+		key != "" &&
+		cluster.Spec.CloudStorage.NodePoolLabel != key {
+		return fmt.Errorf("node pool label key incorrect, expected %s, actual %s", key, cluster.Spec.CloudStorage.NodePoolLabel)
+	}
+
+	return nil
+}
+
+func (c *Controller) getCloudStorageLabelKey(cluster *corev1.StorageCluster) (string, error) {
+	if len(cluster.Spec.Nodes) == 0 {
+		return "", nil
+	}
+
+	key := ""
+	for _, node := range cluster.Spec.Nodes {
+		if node.CloudStorage == nil {
+			continue
+		}
+
+		keys := getKeysFromNodeSelector(&node.Selector)
```

Comment:
So, in case of nodeName selector, we are going to not enforce the validation?

Reply:
i will add it back later. thinking of only enforcing the validation on fresh installation, the problem is the first time we know status is empty it is fresh installation we can fail the validation then update the status in storagecluster, but in second reconcile loop the status is updated so we no longer know it is fresh installation.

Reply:
resolved. also add check to skip validation on existing cluster.